orca-bakery/src/main/groovy/com/netflix/spinnaker/orca/bakery/pipeline/BakeStage.groovy

```diff
@@ -129,7 +136,28 @@ class BakeStage implements StageDefinitionBuilder {
           return deploymentDetails
         }
       ]
+
+      // TODO: add condition for user-specified opt-out setting in addition to feature toggle
+      if (failOnImageNameMismatch()) {
+        def imageNames = globalContext.deploymentDetails.stream().map({details -> details['imageName']})
+          .distinct()
+          .collect(Collectors.toList())
+
+        if (imageNames.size() > 1) {
+          throw new IllegalStateException("Non-matching image names found in different regions: ${imageNames}. " +
+          "Failing bake stage as this would cause deployment failures. Please re-run the bake.")
+        }
+      }
+
```

Comment:
I'd maybe write this more like...

1. Use `Joiner` to create the `imageNames` variable: https://guava.dev/releases/16.0/api/docs/com/google/common/base/Joiner.html
2. Use `com.netflix.spinnaker.kork.exceptions` classes instead of root exceptions.
3. The exception message could be clearer. Perhaps something like this? `Image names found in different regions do not match: $imageNames. Re-run the bake to protect against deployment failures.`

Reply:
Done. For #2, used `ConstraintViolationException`, which was the closest thing I could find. Let me know if you're OK with that or if we should add a new exception type to kork.